client/src/language/codeActionProvider.ts

```diff
@@ -0,0 +1,101 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) 2023 Savoir-faire Linux. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import vscode from 'vscode'
+import { requestsManager } from './RequestManager'
+import { embeddedLanguageDocsManager } from './EmbeddedLanguageDocsManager'
+import { getEmbeddedLanguageDocRange, getIndentationOnLine, getOriginalDocPosition } from './utils'
+import { logger } from '../lib/src/utils/OutputLogger'
+import { type Range } from 'vscode-languageclient'
+
+export class BitbakeCodeActionProvider implements vscode.CodeActionProvider {
+  async provideCodeActions (document: vscode.TextDocument, range: vscode.Range | vscode.Selection, context: vscode.CodeActionContext, token: vscode.CancellationToken): Promise<vscode.CodeAction[]> {
+    const diagnostics = context.diagnostics
+    const actions = await Promise.all(
+      diagnostics.map(async (diagnostic) => await buildActionFromDiagnostic(document, diagnostic))
+    ).then((results) => results.flat())
+    return actions
+  }
+}
+
+const buildActionFromDiagnostic = async (document: vscode.TextDocument, diagnostic: vscode.Diagnostic): Promise<vscode.CodeAction[]> => {
+  const originalRange = diagnostic.range
+  const embeddedLanguageType = await requestsManager.getEmbeddedLanguageTypeOnPosition(document.uri.toString(), originalRange.start)
+  if (embeddedLanguageType === undefined || embeddedLanguageType === null) {
+    return []
+  }
+  const embeddedLanguageDocInfos = embeddedLanguageDocsManager.getEmbeddedLanguageDocInfos(document.uri.toString(), embeddedLanguageType)
+  if (embeddedLanguageDocInfos === undefined || embeddedLanguageDocInfos === null) {
+    return []
+  }
+  const embeddedLanguageTextDocument = await vscode.workspace.openTextDocument(embeddedLanguageDocInfos.uri)
+  const embeddedRange = getEmbeddedLanguageDocRange(
+    document,
+    embeddedLanguageTextDocument,
+    embeddedLanguageDocInfos.characterIndexes,
+    originalRange
+  )
+
+  const tempActions = await vscode.commands.executeCommand<vscode.CodeAction[]>(
+    'vscode.executeCodeActionProvider',
+    embeddedLanguageDocInfos.uri,
+    embeddedRange
+  )
+
+  const actions: vscode.CodeAction[] = []
+  tempActions.forEach((action) => {
+    switch (action.command?.command) {
+      case 'python.addImport':
```

Comment:
Can't we import all action commands in a generic way? Would most of them be broken?

Reply:
In my understanding, there are two ways with code actions to define the code to modify: with a WorkspaceEdit or a Command.

If vscode-python was using WorkspaceEdit, it would be possible to handle it in a generic way, since that object has a clearly defined structure. However, vscode-python uses Command, and those are called with an arbitrary array of arguments.

I guess commands are more flexible, and their result does not have to be computed beforehand. For our needs, I simply convert the Command to a WorkspaceEdit.